index.js

```diff
@@ -498,37 +537,49 @@ MediaSession.prototype = extend(MediaSession.prototype, {
-        var desc = oldLocalDescription;
 
-        // filter to only sources that changed from recvonly to sendrecv
-        desc.contents = desc.contents.filter(function(oldContent) {
-            if (oldContent.application.applicationType === 'rtp' && oldContent.application.sources && oldContent.application.sources.length) {
-                var newContent = findMatchingContentBlock(oldContent, newLocalDescription);
-                // the filter function handles the case where oldContent is null
-                return filterToMatchingRecvonly(oldContent, newContent);
-            }
-        });
-        delete desc.groups;
-        if (desc.contents.length > 0) {
-            this.send('source-remove', desc);
-        }
+
+    // Justin's new functions
+    _determineDifferencesAndSignal: function(oldLocalDescription, newLocalDescription) {
+
+
+        const oldDescCopy = JSON.parse(JSON.stringify(oldLocalDescription));
+        const newDescCopy =  JSON.parse(JSON.stringify(newLocalDescription));
+
+        var diffObject = generateDifferenceOfSources(oldDescCopy, newDescCopy);
+
+        this._signalDifferenceiInSources({sourcesRemoved: diffObject.sourcesRemoved, sourcesAdded: diffObject.sourcesAdded,  sourcesModified: diffObject.sourcesModified,
+             oldLocalDescription: oldDescCopy, newLocalDescription: newDescCopy });
     },
 
-    _addRecvOnlySourceIfNotPresent: function(oldLocalDescription, newLocalDescription) {
-        var desc = newLocalDescription;
+    _signalDifferenceiInSources: function(diffObject){
+        const oldContents = diffObject.oldLocalDescription.contents.filter(getContent);
+        const newContents = diffObject.newLocalDescription.contents.filter(getContent);
+        delete diffObject.oldLocalDescription.groups;
+        delete diffObject.newLocalDescription.groups;
 
-        // filter to only sources that changed from recvonly to sendrecv
-        desc.contents = desc.contents.filter(function(newContent) {
-            if (newContent.application.applicationType === 'rtp' && newContent.application.sources && newContent.application.sources.length) {
-                var oldContent = findMatchingContentBlock(newContent, oldLocalDescription);
-                // the filter function handles the case where oldContent is null
-                return filterToMatchingRecvonly(newContent, oldContent);
-            }
-        });
-        delete desc.groups;
-        if (desc.contents.length > 0) {
-            this.send('source-add', desc);
+        if (diffObject.sourcesAdded.length) {
+            diffObject.newLocalDescription.contents = getProperSSRCS(newContents, diffObject.sourcesAdded);
+            this._log('info', 'sending source add', diffObject.newLocalDescription);
+            this.send('source-add', diffObject.newLocalDescription);
+        }
+
+        if (diffObject.sourcesRemoved.length && !diffObject.sourcesModified.length) { // to avoid signaling remove twice
```

Comment:
Do we need to specifically guard against this case? A single source should never be in both lists, right? Also, you could technically have a source that was changed and a source that was removed

Reply:
good catch.